### PR TITLE
network: fixes to public address support

### LINKED ISSUE #5713: Incorrect Remote Address in wsNetwork logs when algod behind proxy
### Subject of the issue

After the websocket upgrade the value logged for "remote" is remote TCP address instead of the value of header defined in UseXForwardedForAddressField .

in wsNetwork.go

`With("remote", trackedRequest.otherPublicAddr) `
`With("remote", peer.rootURL)`

both log remote TCP address, not the one reported by proxy.

This can be fixed with `trackedRequest.otherPublicAddr = request.RemoteAddr` but affects peer.rootURL and isConnectedTo(addr)

So maybe this should be changed to peer.OriginAddress() - same as in telemetry.

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

1. requestTracker changes:
  - Remove `http.Request.RemoteAddr` overwriting in request tracker
  - Remove `http.Request` from request tracker
  - Add a new r`emoteAddresss()` method that returns a most suitable address

2. public addr changes:
  - no `otherPublicAddr` outside of requestTracker
  - new `remoteAddresss()` method provides most meaningful address for incoming requests

Rationale:
There is a chain http.Handler's in wsNetwork: request tracker and wsNetwork itself.
Tracked request is created/updated in the first `ServeHTTP` and used to save a pointer to `http.Request` object.
This request object is actually gets copied in `mux.ServeHTTP` that calls downstream wsNetwork's `ServeHTTP` making
the request non-usable in the main wsNetwork's ServeHTTP so removed.

`otherPublicAddr` is an address reported by a peer via our custom `X-Algorand-Location` header that the only
correct value for wsPeer.rootURL (host:port accepting connections) but cannot be trusted.
Unconditionally rewriting it with remoteAddr obtained from a tcp connection is too much and a new
`remoteAddresss()` method attempts to provide the most meaningful trusted value.

Fixes #5713

## Test Plan

1. Existing test must pass
2. Added a new unit test
3. Updated some connection dedup `TestPeering` tests since they relied on a buggy `otherPublicAddr` overwriting by `remoteAddr`.

### CODE REVIEW

File: network/requestTracker.go [hunk 1]
```diff
@@ -65,6 +77,19 @@ func makeTrackerRequest(remoteAddr, remoteHost, remotePort string, createTime ti
 	}
 }
 
+// publicAddress returns either otherPublicAddr or possible resloved remoteHost.
```
Comment (algorandskiy) [inline network/requestTracker.go:80]:
```suggestion
// publicAddress returns either otherPublicAddr or possible resloved remoteHost.
```

File: network/requestTracker.go [hunk 2]
```diff
@@ -37,15 +37,27 @@ const (
...
+	created time.Time
+	// remoteHost is IP address of the remote host and it is equal to either
+	// a host part of the remoteAddr or to the value of X-Forwarded-For header.
+	remoteHost string
+	// remotePort is the port of the remote peer as reported by the connection or
+	// by the standard http.Request.RemoteAddr field.
+	remotePort string
+	// remoteAddr is IP:Port of the remote host retrieved from the connection
+	// or from the standard http.Request.RemoteAddr field.
+	// This field the real address of the remote incomming connection.
```
Comment (gmalouf) [inline network/requestTracker.go:52]:
```suggestion
	// This field the real address of the remote incoming connection.
```

File: network/requestTracker.go [hunk 3]
```diff
@@ -37,15 +37,27 @@ const (
...
 type TrackerRequest struct {
-	created            time.Time
-	remoteHost         string
-	remotePort         string
-	remoteAddr         string
-	request            *http.Request
+	created time.Time
+	// remoteHost is IP address of the remote host and it is equal to either
+	// a host part of the remoteAddr or to the value of X-Forwarded-For header.
+	remoteHost string
```
Comment (gmalouf) [inline network/requestTracker.go:46]: Do we want to reference the config parameter that determines this here?

File: network/requestTracker.go [hunk 4]
```diff
@@ -65,6 +77,33 @@ func makeTrackerRequest(remoteAddr, remoteHost, remotePort string, createTime ti
...
+//   - remoteAddr is used otherwise.
+func (tr *TrackerRequest) remoteAddress() string {
+	if len(tr.otherPublicAddr) != 0 {
+		url, err := ParseHostOrURL(tr.otherPublicAddr)
+		if err == nil && len(tr.remoteHost) > 0 && url.Hostname() == tr.remoteHost {
+			return tr.otherPublicAddr
+		}
+	}
+	url, err := ParseHostOrURL(tr.remoteAddr)
+	if err != nil || url.Hostname() == tr.remoteHost {
```
Comment (gmalouf) [inline network/requestTracker.go:100]: Am I reading this right or is the remoteHost and remoteAddr in the description the inverse of how they are used in this section?

Comment (algorandskiy) [reply network/requestTracker.go:100]: The idea is the following: `remoteAddr` is source of truth and a default value. The `url.Hostname() == tr.remoteHost` handles the case with no proxy and `tr.remoteAddr` has the port number, and `remoteHost` does not.

File: network/wsNetwork.go [hunk 5]
```diff
@@ -1025,7 +1025,7 @@ func (wn *WebsocketNetwork) ServeHTTP(response http.ResponseWriter, request *htt
 
 	matchingVersion, otherVersion := wn.checkProtocolVersionMatch(request.Header)
 	if matchingVersion == "" {
-		wn.log.Info(filterASCII(fmt.Sprintf("new peer %s version mismatch, mine=%v theirs=%s, headers %#v", request.RemoteAddr, wn.supportedProtocolVersions, otherVersion, request.Header)))
+		wn.log.Info(filterASCII(fmt.Sprintf("new peer %s version mismatch, mine=%v theirs=%s, headers %#v", trackedRequest.remoteHost, wn.supportedProtocolVersions, otherVersion, request.Header)))
```
Comment (gmalouf) [inline network/wsNetwork.go:1028]: why not `trackedRequest.RemoteAddress()` here?

File: network/wsNetwork.go [hunk 6]
```diff
@@ -1097,7 +1094,7 @@ func (wn *WebsocketNetwork) ServeHTTP(response http.ResponseWriter, request *htt
 	peer.TelemetryGUID = trackedRequest.otherTelemetryGUID
 	peer.init(wn.config, wn.outgoingMessagesBufferSize)
 	wn.addPeer(peer)
-	wn.log.With("event", "ConnectedIn").With("remote", trackedRequest.otherPublicAddr).With("local", localAddr).Infof("Accepted incoming connection from peer %s", trackedRequest.otherPublicAddr)
+	wn.log.With("event", "ConnectedIn").With("remote", trackedRequest.remoteAddress()).With("local", localAddr).Infof("Accepted incoming connection from peer %s", trackedRequest.remoteAddr)
```
Comment (gmalouf) [inline network/wsNetwork.go:1097]: Similar ask, do the two different calls make sense here?

File: network/wsNetwork_test.go [hunk 7]
```diff
@@ -1432,13 +1424,16 @@ func TestPeeringWithIdentityChallenge(t *testing.T) {
...
 
+	assert.Equal(t, 1, len(netA.GetPeers(PeersConnectedOut)))
 	// A connects to C (but uses addrB here to simulate case where B & C have the same PublicAddress)
 	netA.wg.Add(1)
 	netA.tryConnect(addrB, gossipC)
 	// let the tryConnect go forward
 	time.Sleep(250 * time.Millisecond)
+	assert.Eventually(t, func() bool {
+		return len(netA.GetPeers(PeersConnectedOut)) == 2
+	}, time.Second, 50*time.Millisecond)
```
Comment (gmalouf) [inline network/wsNetwork_test.go:1436]: This time window might be a risk for the build server

Comment (algorandskiy) [reply network/wsNetwork_test.go:1436]: actually, good catch, I need to remove the `time.Sleep(250 * time.Millisecond)` above and leave only this retrying logic. Note, 1 sec is greater than the 250ms timeout before.

File: network/requestTracker.go (hunk 3 above)
Comment (algorandskiy) [reply network/requestTracker.go:46]: added

File: network/wsNetwork_test.go [hunk 8]
```diff
@@ -1381,25 +1375,25 @@ func TestPeeringWithIdentityChallenge(t *testing.T) {
...
-		netA.tryConnect(addrB, gossipB)
-		// let the tryConnect go forward
-		time.Sleep(250 * time.Millisecond)
-	}
+	_, ok = netA.tryConnectReserveAddr(addrB)
+	assert.False(t, ok)
+	netA.wg.Add(1)
+	netA.tryConnect(addrB, gossipB)
+	// let the tryConnect go forward
+	time.Sleep(250 * time.Millisecond)
```
Comment (gmalouf) [inline network/wsNetwork_test.go:1393]: Can we do this without a single sleep - perhaps periodic waitUntil type checks with a long interruptible timeout on the checking? Getting these sleep times right is tricky (also slows the build down) and can lead to random failures when the build server is congested.

Comment (gmalouf) [reply network/wsNetwork_test.go:1393]: assert.Eventually is what I am thinking of

File: network/wsNetwork_test.go [hunk 9]
```diff
@@ -1411,11 +1405,9 @@ func TestPeeringWithIdentityChallenge(t *testing.T) {
...
 	// on the peer will detect and close the peer. Since this is asynchronous,
 	// we wait and check regularly to allow the connection to settle
-	assert.Eventually(
-		t,
-		func() bool { return len(netB.GetPeers(PeersConnectedIn)) == 1 },
-		5*time.Second,
-		100*time.Millisecond)
+	assert.Eventually(t, func() bool {
+		return len(netB.GetPeers(PeersConnectedIn)) == 1
+	}, time.Second, 50*time.Millisecond)
```
Comment (gmalouf) [inline network/wsNetwork_test.go:1410]: is this timeout long enough on a build server? (Eventually is checking constantly, so this isn't a minimum duration, just a maximum)

Comment (gmalouf) [reply network/wsNetwork_test.go:1410]: Answered below, all set.

File: network/wsNetwork_test.go (hunk 8 above)
Comment (algorandskiy) [reply network/wsNetwork_test.go:1393]: idk, this connection attempt is going to be aborted and `tryConnect` does not return an error. I'll check if I can rely on a counter or a logged message here

File: network/requestTracker.go (hunk 4 above)
Comment (algorandskiy) [reply network/requestTracker.go]: updated

**Review by jasonpaulos — DISMISSED** [review]
This seems better. Just putting in some minor corrections

**Review by jasonpaulos — APPROVED** [review]

**Pull request merged**